Source/Core/VideoCommon/PixelShaderGen.cpp

```diff
@@ -1336,15 +1336,27 @@ ShaderCode GeneratePixelShaderCode(APIType api_type, const ShaderHostConfig& hos
       {
         out.Write("\t\tcustom_data.final_color = float4(ocol0.x / 255.0, ocol0.y / 255.0, ocol0.z "
                   "/ 255.0, ocol0.w / 255.0);\n");
-        out.Write("\t\tfloat3 custom_output = {}_{}(custom_data).xyz;\n",
-                  CUSTOM_PIXELSHADER_COLOR_FUNC, i);
-        out.Write("\t\tocol0.xyz = uint3(custom_output.x * 255, custom_output.y * 255, "
-                  "custom_output.z * 255);\n");
+        out.Write("\t\tfloat4 custom_output = {}_{}(custom_data);\n", CUSTOM_PIXELSHADER_COLOR_FUNC,
+                  i);
+        out.Write("\t\tocol0 = uint3(custom_output.x * 255, custom_output.y * 255, "
+                  "custom_output.z * 255, custom_output.w * 255);\n");
+
+        if (use_dual_source)
+        {
+          out.Write("\t\tocol1.w = ocol0.w;\n");
+          out.Write("\t\tocol0.w = 255;\n");
+        }
       }
       else
       {
         out.Write("\t\tcustom_data.final_color = ocol0;\n");
-        out.Write("\t\tocol0.xyz = {}_{}(custom_data).xyz;\n", CUSTOM_PIXELSHADER_COLOR_FUNC, i);
+        out.Write("\t\tocol0 = {}_{}(custom_data);\n", CUSTOM_PIXELSHADER_COLOR_FUNC, i);
+
+        if (use_dual_source)
+        {
+          out.Write("\t\tocol1.w = ocol0.w;\n");
+          out.Write("\t\tocol0.w = 1.0;\n");
+        }
```

Comment:
This logic doesn't seem to be correct. It should match this:

https://github.com/dolphin-emu/dolphin/blob/59911b0668be6124abec623fe018252e909c0180/Source/Core/VideoCommon/PixelShaderGen.cpp#L2109-L2125

i.e. use `I_ALPHA` if `uid_data->useDstAlpha` otherwise `prev.a`. I'm not entirely sure where this would show up, though.

It might be easier to move the custom shader stuff before `WriteColor`, and modify `prev` instead of `ocol0`.

Reply:
I don't want to use pixel shader's values though.  The idea is to use the custom value (4th value) from the custom shader?

Also, can't put it before because the color is expected to overwrite the color in the game.

Reply:
So if the user has a shader that returns `float4(1, 1, 1, 0.5)`, they would get a white pixel that was 50% transparent (showing the background).

Reply:
You probably want to still use `I_ALPHA` if `useDstAlpha` is true, but right, you'd want to use alpha from the custom value, not `prev.a`.

I'm pretty sure you can move it to before `WriteColor` though; you'd just want to modify `prev` with the custom shader's output instead of modifying `ocol0` directly, and then `WriteColor` would copy into `ocol0` and/or `ocol1` as needed.

Reply:
> You probably want to still use I_ALPHA if useDstAlpha is true

But `I_ALPHA` comes from the game doesn't it?  That would mean the user has no control over the resulting image in that case.  It's not super obvious what `I_ALPHA` is though.  Looks like a constant passed in.

Reply:
Yeah, it comes from the game (specifically `bpmem.constantalpha`).

If I recall correctly, it's generally used for using the current alpha value in the framebuffer for blending while also resetting the alpha value in the framebuffer to a known constant value afterwards. I believe Super Mario Galaxy uses it for shadows, but I haven't looked into it in a while. This seems like behavior that you'd want to preserve with a custom shader.

Reply:
Thanks Pokechu, yeah, I found it here:  https://github.com/dolphin-emu/dolphin/blob/d3cf35019c0a201f201f2a80bf4354034a944dd6/Source/Core/VideoCommon/PixelShaderManager.cpp#L239

Well, the custom shader is meant to completely overwrite the game's rendering.  And there's not really a concept of destination or src alpha, in a modern shader pipeline... just an alpha value.  If the user wanted those concepts, they could apply it in the shader with their own values (or using the TEV input).

I did test the destination alpha on Xenoblade's torch particle effects and setting the alpha value blended the particles more or less as expected (using the game's final color displayed the effect as if no custom shader was present).

I own SMG2 and could take a look at its shadows.